src/Control/Effect/Cull.hs

```diff
@@ -0,0 +1,60 @@
+{-# LANGUAGE DeriveFunctor, ExistentialQuantification, FlexibleContexts, FlexibleInstances, LambdaCase, MultiParamTypeClasses, StandaloneDeriving, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Cull where
+
+import Control.Effect.Carrier
+import Control.Effect.Internal
+import Control.Effect.NonDet
+import Control.Effect.Sum
+
+-- | 'Cull' effects are used with 'NonDet' to provide control over branching.
+data Cull m k
+  = forall a . Cull (m a) (a -> k)
+
+deriving instance Functor (Cull m)
+
+instance HFunctor Cull where
+  hmap f (Cull m k) = Cull (f m) k
+  {-# INLINE hmap #-}
+
+instance Effect Cull where
+  handle state handler (Cull m k) = Cull (handler (m <$ state)) (handler . fmap k)
+  {-# INLINE handle #-}
+
+-- | Cull nondeterminism in the argument, returning at most one result.
+--
+--   prop> run (runNonDet (runCull (cull (pure a <|> pure b)))) == [a]
+--   prop> run (runNonDet (runCull (cull (pure a <|> pure b) <|> pure c))) == [a, c]
+--   prop> run (runNonDet (runCull (cull (asum (map pure (repeat a)))))) == [a]
+cull :: (Carrier sig m, Member Cull sig) => m a -> m a
```

Comment:
This hints that we may be able to define other search heuristics as scoped effects. See #69 for more discussion.